rules/onchain_budget.go

```diff
@@ -0,0 +1,687 @@
+	}
+
+	if !checker.HandlesRequest(msg.ProtoReflect().Type()) {
+		return nil, fmt.Errorf("invalid implementation, checker "+
+			"for URI %s does not accept request of type %v",
+			uri, msg.ProtoReflect().Type())
+	}
+
+	return checker.HandleRequest(ctx, msg)
+}
+
+// HandleResponse handles and possible alters a response.
+//
+// NOTE: this is part of the Rule interface.
+func (o *OnChainBudgetEnforcer) HandleResponse(ctx context.Context, uri string,
+	msg proto.Message) (proto.Message, error) {
+
+	o.Lock()
+	defer o.Unlock()
+
+	checkers := o.checkers()
+	if checkers == nil {
+		return nil, nil
+	}
+
+	checker, ok := checkers[uri]
+	if !ok {
+		return nil, nil
+	}
+
+	if !checker.HandlesResponse(msg.ProtoReflect().Type()) {
+		return nil, fmt.Errorf("invalid implementation, checker for "+
+			"URI %s does not accept response of type %v", uri,
+			msg.ProtoReflect().Type())
+	}
+
+	return checker.HandleResponse(ctx, msg)
+}
+
+// HandleErrorResponse handles and possible alters an error. This is a noop for
+// the OnChainBudgetEnforcer rule.
+//
+// NOTE: this is part of the Enforcer interface.
+func (o *OnChainBudgetEnforcer) HandleErrorResponse(ctx context.Context,
+	uri string, respErr error) (error, error) {
+
+	o.Lock()
+	defer o.Unlock()
+
+	checkers := o.checkers()
+	if checkers == nil {
+		return nil, nil
+	}
+
+	checker, ok := checkers[uri]
+	if !ok {
+		return nil, nil
+	}
+
+	return checker.HandleErrorResponse(ctx, respErr)
+}
+
+// checkers returns a map of URI to rpcmiddleware.RoundTripChecker which define
+// how the URI should be handled.
+func (o *OnChainBudgetEnforcer) checkers() map[string]mid.RoundTripChecker {
+	// handleOpenChannelRequest is a helper function which handles the
+	// request for opening a single channel.
+	handleOpenChannelRequest := func(ctx context.Context,
+		r *lnrpc.OpenChannelRequest) (proto.Message, error) {
+
+		if r.FundMax {
+			return nil, fmt.Errorf("max funding not supported")
+		}
+
+		if len(r.Outpoints) > 0 {
+			return nil, fmt.Errorf("outpoints not supported")
+		}
+
+		if r.TargetConf != 0 {
+			return nil, fmt.Errorf("target conf not supported")
+		}
+
+		err := o.checkFeeRate(r.SatPerVbyte)
+		if err != nil {
+			return nil, err
+		}
+
+		// Prepend an identifier to later be able to check the action's
+		// outcome. This is chosen randomly to avoid any potential
+		// collisions with user provided memos.
+		id := randId(idLen)
+		r.Memo = fmt.Sprintf("%s:%s", id, r.Memo)
+
+		amt := uint64(r.LocalFundingAmount + r.PushSat)
+		err = o.handlePendingPayment(
+			ctx,
+			&onChainAction{
+				Amount:     amt,
+				ActionType: actionTypeOpenChannel,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		return r, nil
+	}
+
+	// handleOpenChannelError is a helper function which handles the error
+	// returned for all the channel opening endpoints.
+	handleOpenChannelError := func(ctx context.Context, _ error) (error,
+		error) {
+
+		return nil, o.cancelPendingPayment(ctx)
```

Comment:
This shouldn't return `nil` for the first return value, right? It should be set to the `error` passed as the second arg to the function, given that `o.cancelPendingPayment(ctx)` doesn't error.

Reply:
right, great catch! :pray: